File: CSpack/CSpack/CSpack/solvers/NL0R.m
function out = NL0R(data,n,lambda,pars)
% This code aims at solving the L0 regularized optimization with form
% 
%         min_{x\in R^n} 0.5||Ax-b||^2 + \lambda \|x\|_0 
% 
% where \lambda is updated iteratively.
%       A\in\R{m by n} the measurement matrix
%       b\in\R{m by 1} the observation vector 
% =========================================================================
% Inputs:
%     data:     A structure (required)
%               (data.A, data.b) if A is a matrix 
%               (data.A, data.b, data.At) if A is a function handle
%               i.e., Ax = data.A(x); A'y = data.At(y); 
%     n:        Dimension of the solution x, (required)             
%     pars:     Parameters are all OPTIONAL
%               pars.x0      --  Starting point of x (default, zeros(n,1))
%               pars.tau     --  A positive scalar (default,(n<=1e3)+(n>1e3)/2)
%               pars.lam     --  An initial penalty parameter (default, maxlam/2)
%               pars.rate    --  A positive scalar to adjust lam, (default, rate0) 
%               pars.disp    --  Display or not results for each iteration (default, 1) 
%               pars.maxit   --  Maximum number of iterations, (default,2000) 
%               pars.tol     --  Tolerance of the halting condition, (default,1e-6)
%               pars.obj     --  A predefined lower bound of f(x), (default,1e-20)
% Outputs:
%     out.sol:           The sparse solution x
%     out.sp:            Sparsity level of out.sol 
%     out.time           CPU time
%     out.iter:          Number of iterations
%     out.obj:           Objective function value at out.sol 
% ========================================================================= 
% This code is programmed based on the algorithm proposed in 
% S. Zhou, L. Pan and N. Xiu, Numerical Algorithms, 2020, 
% Newton Method for l_0 Regularized Optimization.
% Send your comments and suggestions to <slzhou2021@163.com> 
% Warning: Accuracy may not be guaranteed !!!!! 
% ========================================================================= 

warning off;

t0 = tic;
if nargin<2
   fprintf(' No enough inputs. No problems will be solverd!'); return;
end

if ~isfield(data,'A')
    fprintf('<data.A> is missing, unable to run the solver ...');
    return
else
    if  isa(data.A,'function_handle') && ~isfield(data,'At') 
        fprintf('<data.At> is missing, unable to run the solver ...');
        return
    end
end
if  ~isfield(data,'b')
    fprintf('<data.b> is missing, unable to run the solver ...');
    return
end

if isstruct(data);  data.n = n; end
func   = @(x,T1,T2)CS(x,T1,T2,data);

if nargin<3; pars=[]; end    
    
rate0 = (n<=1e3)*0.5+(n>1e3)/exp(3/log10(n));
tau0  = (n<=1e3)+(n>1e3)/2; 
if isfield(pars,'x0');    x0    = pars.x0;    else; x0 = zeros(n,1);end
if isfield(pars,'tau');   tau   = pars.tau;   else; tau   = tau0;   end
if isfield(pars,'rate');  rate  = pars.rate;  else; rate  = rate0;  end
if isfield(pars,'disp');  disp  = pars.disp;  else; disp  = 1;      end
if isfield(pars,'maxit'); itmax = pars.maxit; else; itmax = 2000;   end
if isfield(pars,'obj');   pobj  = pars.obj;   else; pobj  = 1e-30;  end 
if isfield(pars,'tol');   tol   = pars.tol;   else; tol   = 1e-10;  end 


x       = x0;
Err     = zeros(1,itmax);
Obj     = zeros(1,itmax);
Nzx     = zeros(1,itmax);
FNorm   = @(x)norm(x)^2;

if disp  
    fprintf(' \n Start to run the solver -- NL0R\n');
    fprintf(' -------------------------------------\n');
    fprintf(' Iter     ObjVal    Sparsity     Time \n'); 
    fprintf(' -------------------------------------\n');
end

% Initial check for the starting point
[obj,g] = func(x,[],[]);
if FNorm(g)==0 
   fprintf('Starting point is a good stationary point, stop !!!\n'); 
   out.sol = x;
   out.obj = obj;
   return;
else   
   maxlam = max(abs(g))^2*tau/2;
end

if  max(isnan(g))
    x       = zeros(n,1);
    rind    = randi(n);
    x(rind) = rand;
    [obj,g] = func(x,[],[]);
end

pcgit   = 5;
pcgtol  = 1e-5;
beta    = 0.5;
sigma   = 5e-5;
delta   = 1e-10;
T0      = [];  
nx      = 0;

% The main body  
for iter  = 1:itmax
    x0    = x;  
    xtg   =  x0-tau*g ; 
    T     = find(abs(xtg)>sqrt(2*tau*lambda));     
    nT    = nnz(T);
    if nT > max(0.12,0.2/log2(1+iter))*n 
       Tnew  = SparseApprox(xtg(T),T); 
       nTnew = nnz(Tnew );
       if ~isempty(Tnew) && nT/nTnew < 20
           T  = Tnew;  
           nT = nTnew;
       end 
    end
    TTc   = setdiff(T0,T);
    flag  = isempty(TTc);    

    % Calculate the error for stopping criteria 
    FxT       = sqrt(FNorm(g(T))+FNorm(x(TTc)));
    Err(iter) = FxT/sqrt(n); 
    Nzx(iter) = nx;
    if  disp  
         fprintf('%4d     %5.2e    %4d    %6.2fsec\n',iter, obj, nT, toc(t0)); 
    end
    
    % Stopping criteria   

    stop0  = iter>1 && abs(obj-obj0)<1e-6*(1+obj);
    stop1  = Err(iter)<tol && nx==nT && stop0 && flag;  
    stop2  = iter > 3 && obj < pobj && nx<=ceil(n/4);
    stop3  = iter > 9 && std(Nzx(iter-9:iter))<= 0 &&...
             std(Err(iter-9:iter))^2 <= min(Err(iter-9:iter)) &&...
             std(Obj(iter-9:iter))^2 <= min(Obj(iter-9:iter-1)); 
    stop4  = norm(g)<tol && nx<=ceil(n/4);      
    if stop1 || stop2 || stop3 || stop4; break;   end
   
    % update next iterate
    if  iter   == 1 || flag    % two consective iterates have same supports
        H       = func(x0,T,[]);     
        if isa(H,'function_handle')
           d    = my_cg(H,-g(T),pcgtol,pcgit,zeros(nT,1)); 
        else 
           d    = -H\g(T);  
        end
       
        dg     = sum(d.*g(T));
        ngT    = FNorm(g(T));
        if dg  > max(-delta*FNorm(d), -ngT) || isnan(dg) 
        d      = -g(T); 
        dg     = ngT; 
        end
    else                  % two consective iterates have different supports                       
        [H,D]   = func(x0,T,TTc);             
        if  isa(D,'function_handle')
            rhs = D(x0(TTc))-g(T);  
        else
            rhs = D*x0(TTc) - g(T); 
        end
        if  isa(H,'function_handle')
            d   = my_cg(H, rhs,pcgtol,pcgit,zeros(nT,1));  
        else 
            d   = H\rhs; 
        end
         
        Fnz    = FNorm(x(TTc))/4/tau;
        dgT    = sum(d.*g(T));
        dg     = dgT-sum(x0(TTc).*g(TTc));
        
        delta0 = delta;
        if Fnz > 1e-4; delta0 = 1e-4; end
 
        ngT    = FNorm(g(T));
        if dgT > max(-delta0*FNorm(d)+Fnz, -ngT) || isnan(dg) 
           d   = -g(T); 
           dg  = ngT; 
        end            
    end
    
    % Armijo line search
    alpha    = 1; 
    x        = zeros(n,1);    
    obj0     = obj;             
    for i      = 1:6
        x(T)   = x0(T) + alpha*d;
        obj    =  func(x,[],[]);
        if obj < obj0  + alpha*sigma*dg; break; end        
        alpha  = beta*alpha;
    end
    
    T0       = T; 
    [obj,g]  = func(x,[],[]);
    Obj(iter)= obj; 
    
%   Update tau    
    if  mod(iter,10)==0  
        OBJ = Obj(iter-9:iter);
        if Err(iter)>1/iter^2 || sum(OBJ(2:end)>1.5*OBJ(1:end-1))>=2 
            if iter<1500; tau = tau/1.25; 
            else;         tau = tau/1.5; 
            end     
        else          
            tau = tau*1.25;   
        end
    end 
    
%   Update lambda    
    nx  = nnz(x); 
    if  iter>5 && (nx > 2*max(Nzx(1:iter-1))) && Err(iter)<1e-2
        rate0   = 2/rate;   
        x       = x0;
        nx      = nnz(x0); 
        nx0     = Nzx(iter-1);  
        [obj,g] = func(x,[],[]);
        rate    = 1.1;
    else  
        rate0   = rate;
    end
       
    if exist('nx0') && nx < nx0
       rate0 = 1;   
    end
 
    if mod(iter,1)==0 
       lambda  = min(maxlam,lambda*(2*(nx>=0.1*n)+rate0)); 
    end
    
end

%Results output ------------------------------------------------- 
sol         = zeros(n,1); 
sol(T)      = x(T);
iter        = iter-1;
[obj,g]     = func(sol ,[],[]);
time        = toc(t0);
out.sp      = nnz(sol );  
out.time    = time;
out.iter    = iter;
out.sol     = sol;
out.obj     = obj;   
out.error   = FNorm(g); 
end

% get the sparse approximation of x ---------------------------------------
function T = SparseApprox(x0,T0)
    x      = abs(x0); 
    sx     = sort(x(x~=0));  
    if  length(sx)<=2
        th = sx(end);
    else
        [mx,it] = max(normalize(sx(2:end)./sx(1:end-1)));
        th      = 0; 
        if mx   > 10 && it(1)>1
           th   = sx(it(1)); 
        end
    end       
    T = T0(x>th); 
end

% define functions --------------------------------------------------------
function [out1,out2] = CS(x,T1,T2,data)

if ~isa(data.A, 'function_handle') % A is a matrix 
    if  isempty(T1) && isempty(T2) 
        if  nnz(x) >= 0.8*length(x)
            Axb     = data.A*x-data.b;
        else
            Tx      = find(x); 
            Axb     = data.A(:,Tx)*x(Tx)-data.b;
        end
            out1    = (Axb'*Axb)/2;               % objective function value of f
        if  nargout == 2
            out2    = (Axb'*data.A)';                % gradien of f
        end
    else        
            AT = data.A(:,T1); 
        if  length(T1)<3000
            out1 = AT'*AT;                        %subHessian containing T1 rows and T1 columns
        else
            out1 = @(v)( (AT*v)'*AT )';      
        end       
        if  nargout == 2
            out2 = @(v)( (data.A(:,T2)*v)'*AT )'; %subHessian containing T1 rows and T2 columns
        end       
    end
else  % A is a function handle A*x=A(x)  
    if ~isfield(data,'At') 
        disp('The transpose-data.At-is missing'); return; 
    end
    if ~isfield(data,'n')  
        disp('The dimension-data.n-is missing');  return;  
    end   
    if  isempty(T1) && isempty(T2)  
        Axb  = data.A(x)-data.b;
        out1 = (Axb'*Axb)/2;              % objective function value of f
        if  nargout>1 
            out2 = data.At(Axb);          % gradien of f
        end
    else
        func = fgH(data);    
        out1 = @(v)func(v,T1,T1);         % subHessian containing T1 rows and T1 columns
        if  nargout>1
            out2 = @(v)func(v,T1,T2);     % subHessian containing T1 rows and T1 columns
        end  
        
    end
end

end

function Hess = fgH(data)
    suppz     = @(z,t)supp(data.n,z,t);
    sub       = @(z,t)z(t,:);
    Hess      = @(z,t1,t2)(sub( data.At( data.A(suppz(z,t2))),t1)); 
end

function z = supp(n,x,T)
    z      = zeros(n,1);
    z(T)   = x;
end
% conjugate gradient-------------------------------------------------------
function x = my_cg(fx,b,cgtol,cgit,x)
    if ~isa(fx,'function_handle'); fx = @(v)fx*v; end
    r = b;
    if nnz(x)>0; r = b - fx(x);  end
    e = norm(r,'fro')^2;
    t = e;
    p = r;
    for i = 1:cgit  
        if e < cgtol*t; break; end
        w  = fx(p);
        pw = p.*w;
        a  = e/sum(pw(:));
        x  = x + a * p;
        r  = r - a * w;
        e0 = e;
        e  = norm(r,'fro')^2;
        p  = r + (e/e0)*p;
    end 
end

File: CSpack/CSpack/CSpack/solvers/NL0R.py
import numpy as np
import time
from typing import Callable, Dict, Optional, Tuple, Union

ArrayLike = np.ndarray
FuncType = Callable[[np.ndarray, str, Optional[np.ndarray], Optional[np.ndarray]], Union[Tuple[float, np.ndarray], Tuple[np.ndarray, Optional[np.ndarray]], Callable[[np.ndarray], np.ndarray]]]


def NL0R(func: FuncType, n: int, lam: float, pars: Optional[Dict] = None):
    """Python port of NL0R.m."""
    pars = pars or {}
    x, eta, tol, maxit, disp, update, uppf, rate = setparameters(n, pars)
    x = x.astype(float, copy=True).reshape(n)

    Err = np.zeros(maxit)
    Obj = np.zeros(maxit)
    Nzx = np.zeros(maxit)
    TMP = np.zeros(3)

    def call_fg(vec: np.ndarray) -> Tuple[float, np.ndarray]:
        obj, grad = func(vec, 'fg', None, None)
        grad = np.asarray(grad).reshape(-1)
        return float(obj), grad

    def call_h(vec: np.ndarray, T1: np.ndarray, T2: Optional[np.ndarray]):
        return func(vec, 'h', T1, T2)

    obj, g = call_fg(x)
    if np.linalg.norm(g) == 0:
        if disp:
            print('Starting point is a good stationary point, stop !!!')
        return {
            'sol': x,
            'obj': obj,
            'iter': 0,
            'time': 0.0,
            'sparsity': int(np.count_nonzero(x)),
            'Obj': Obj,
        }

    if np.isnan(g).any():
        x = np.zeros(n)
        rind = np.random.randint(n)
        x[rind] = np.random.rand()
        obj, g = call_fg(x)

    t = 0
    xtg = np.abs(x - eta * g)
    while t < 20:
        T = np.flatnonzero(xtg > np.sqrt(2 * eta * lam))
        nT = T.size
        if nT == 0:
            lam /= 1.25
        elif nT > 0.12 * n:
            lam *= 1.25
        else:
            break
        t += 1

    maxlam = (np.max(np.abs(g)) ** 2) / eta / 2
    nx = 0
    nx0_prev = None
    pcgit = 5
    pcgtol = 1e-5
    beta = 0.5
    sigma = 5e-5
    delta = 1e-10
    T0 = np.array([], dtype=int)

    if disp:
        print(' Start to run the solver -- NL0R ')
        print(' --------------------------------------------------------')
        print('  Iter      Error       Objective    Sparsity    Time(sec)')
        print(' --------------------------------------------------------')

    start = time.perf_counter()
    for iter_idx in range(maxit):
        x0 = x.copy()
        xtg = x0 - eta * g

        while True:
            T = np.flatnonzero(np.abs(xtg) > np.sqrt(2 * eta * lam))
            nT = T.size
            if nT > 0:
                break
            lam /= 1.05

        if iter_idx > 0 and 0 <= nT - T0.size <= 5 and Err[iter_idx - 1] < tol:
            lam = lam0
            while True:
                T = np.flatnonzero(np.abs(xtg) > np.sqrt(2 * eta * lam))
                nT = T.size
                if nT > 0:
                    break
                lam /= 1.05

        if iter_idx >= 0 and nT > max(0.12, 0.2 / np.log2(1 + iter_idx + 1)) * n:
            Tnew = SparseApprox(xtg[T], T)
            nTnew = Tnew.size
            if Tnew.size and nT / nTnew < 20 and nT != nTnew:
                T = Tnew
                nT = nTnew

        TTc = np.setdiff1d(T, T0, assume_unique=False)
        flag = TTc.size == 0

        FxT = np.sqrt(np.linalg.norm(g[T]) ** 2 + np.linalg.norm(x[TTc]) ** 2)
        Err[iter_idx] = FxT / np.sqrt(n)
        Nzx[iter_idx] = nx
        if disp and ((iter_idx + 1) % 10 == 0 or iter_idx + 1 < 100):
            print(f" {iter_idx + 1:4d}      {FxT:8.2e}     {obj:9.2e}      {nx:4d}      {time.perf_counter() - start:5.3f}sec")

        TMP = np.roll(TMP, -1)
        TMP[-1] = obj
        stop1 = Err[iter_idx] < tol and np.std(TMP) < 1e-8 * (1 + abs(obj))
        stop1 = stop1 and nx == nT and flag
        stop2 = (iter_idx + 1) > 3 and obj < uppf and nx <= int(np.ceil(n / 4))
        stop3 = np.linalg.norm(g) < tol and nx <= int(np.ceil(n / 4))
        if iter_idx > 0 and (stop1 or stop2 or stop3):
            if disp and not ((iter_idx + 1) % 10 == 0 or iter_idx + 1 < 100):
                print(f" {iter_idx + 1:4d}      {FxT:8.2e}     {obj:9.2e}      {nx:4d}      {time.perf_counter() - start:5.3f}sec")
            break

        if iter_idx == 0 or flag:
            H = call_h(x0, T, None)
            if callable(H):
                d = my_cg(H, -g[T], pcgtol, pcgit, np.zeros(nT))
            else:
                d = np.linalg.solve(H, -g[T])

            dg = np.dot(d, g[T])
            ngT = np.linalg.norm(g[T]) ** 2
            if dg > max(-delta * np.linalg.norm(d) ** 2, -ngT) or np.isnan(dg):
                d = -g[T]
                dg = ngT
        else:
            res = call_h(x0, T, TTc)
            H, D = res if isinstance(res, tuple) else (res, None)
            rhs = (D(x0[TTc]) if callable(D) else D @ x0[TTc]) - g[T]
            if callable(H):
                d = my_cg(H, rhs, pcgtol, pcgit, np.zeros(nT))
            else:
                d = np.linalg.solve(H, rhs)

            Fnz = np.linalg.norm(x[TTc]) ** 2 / (4 * eta)
            dgT = np.dot(d, g[T])
            dg = dgT - np.dot(x0[TTc], g[TTc])
            delta0 = 1e-4 if Fnz > 1e-4 else delta
            ngT = np.linalg.norm(g[T]) ** 2
            if dgT > max(-delta0 * np.linalg.norm(d) ** 2 + Fnz, -ngT) or np.isnan(dg):
                d = -g[T]
                dg = ngT

        alpha = 1.0
        obj0 = obj
        x_candidate = np.zeros_like(x)
        for _ in range(6):
            x_candidate.fill(0.0)
            x_candidate[T] = x0[T] + alpha * d
            obj_candidate, _ = call_fg(x_candidate)
            if obj_candidate < obj0 + alpha * sigma * dg:
                obj = obj_candidate
                x = x_candidate.copy()
                break
            alpha *= beta
        else:
            x = x_candidate.copy()
            obj, _ = call_fg(x)

        T0 = T.copy()
        obj, g = call_fg(x)
        Obj[iter_idx] = obj

        if (iter_idx + 1) % 10 == 0:
            recent = Obj[max(0, iter_idx - 9):iter_idx + 1]
            if Err[iter_idx] > 1 / (iter_idx + 1) ** 2 or np.sum(recent[1:] > 1.5 * recent[:-1]) >= 2:
                eta = eta / (1.25 if iter_idx + 1 < 1500 else 1.5)
            else:
                eta = eta * 1.25

        nx = int(np.count_nonzero(x))
        if (iter_idx + 1) > 5:
            prev_vals = Nzx[:iter_idx]
            prev_max = np.max(prev_vals) if prev_vals.size else 0
        else:
            prev_max = 0

        if (iter_idx + 1) > 5 and nx > 2 * prev_max and Err[iter_idx] < 1e-2:
            rate0 = 2 / rate
            x = x0
            nx = int(np.count_nonzero(x0))
            nx0_prev = Nzx[iter_idx - 1] if iter_idx > 0 else None
            obj, g = call_fg(x)
            rate = 1.1
        else:
            rate0 = rate

        if nx0_prev is not None and nx < nx0_prev:
            rate0 = 1

        lam0 = lam
        if update:
            lam = min(maxlam, lam * (2 * (nx >= 0.1 * n) + rate0))

    final_obj, g = call_fg(x)
    elapsed = time.perf_counter() - start
    out = {
        'sol': x,
        'obj': final_obj,
        'iter': iter_idx + 1,
        'time': elapsed,
        'sparsity': int(np.count_nonzero(x)),
        'Obj': Obj,
    }

    if disp:
        print(' --------------------------------------------------------')
        normgrad = np.linalg.norm(g)
        if normgrad < 1e-6:
            print(' A global optimal solution might be found')
            print(f' because of ||gradient|| = {normgrad:5.2e}!')
        print(' --------------------------------------------------------')

    return out


def setparameters(n: int, pars: Dict):
    if n <= 1e3:
        rate0 = 0.5
    else:
        rate0 = 1.0 / np.exp(3 / np.log10(n))

    x0 = np.asarray(pars.get('x0', np.zeros(n))).reshape(-1)
    eta = float(pars.get('eta', 1))
    rate = float(pars.get('rate', rate0))
    disp = int(pars.get('disp', 1))
    maxit = int(pars.get('maxit', 2000))
    uppf = float(pars.get('uppf', -np.inf))
    tol = float(pars.get('tol', 1e-6))
    update = int(pars.get('update', 1))

    return x0, eta, tol, maxit, disp, update, uppf, rate


def SparseApprox(x0: np.ndarray, T0: np.ndarray) -> np.ndarray:
    x = np.abs(x0)
    nz = x[x != 0]
    if nz.size == 0:
        return T0
    sx = np.sort(nz)
    if sx.size <= 2:
        th = sx[-1]
    else:
        ratios = sx[1:] / sx[:-1]
        if ratios.size == 0:
            th = sx[-1]
        else:
            mean = ratios.mean()
            std = ratios.std(ddof=0)
            std = std if std != 0 else 1.0
            scores = (ratios - mean) / std
            idx = int(np.argmax(scores))
            th = 0.0
            if scores[idx] > 10 and idx >= 1:
                th = sx[idx]
            else:
                th = sx[-1]
    mask = x > th
    return T0[mask]


def my_cg(fx: Union[Callable[[np.ndarray], np.ndarray], np.ndarray], b: np.ndarray, cgtol: float, cgit: int, x0: np.ndarray) -> np.ndarray:
    bnorm = np.linalg.norm(b)
    if bnorm == 0:
        return np.zeros_like(x0)

    if not callable(fx):
        mat = np.asarray(fx)
        fx = lambda v: mat @ v

    x = x0.copy()
    r = b - fx(x) if np.count_nonzero(x) else b.copy()
    e = np.linalg.norm(r) ** 2
    t = e
    p = r.copy()
    for _ in range(cgit):
        if e < cgtol * t:
            break
        w = fx(p)
        pw = np.dot(p, w)
        if pw == 0:
            break
        a = e / pw
        x = x + a * p
        r = r - a * w
        e0 = e
        e = np.linalg.norm(r) ** 2
        if e0 == 0:
            break
        p = r + (e / e0) * p
    return x
